sessiontxn/isolation/readcommitted.go

```diff
@@ -117,15 +114,30 @@ func (p *PessimisticRCTxnContextProvider) prepareStmtTS() {
 	switch {
 	case p.stmtUseStartTS:
 		stmtTSFuture = sessiontxn.FuncFuture(p.getTxnStartTS)
-	case p.availableRCCheckTS != 0 && sessVars.StmtCtx.RCCheckTS:
-		stmtTSFuture = sessiontxn.ConstantFuture(p.availableRCCheckTS)
+	case p.latestOracleTSValid && sessVars.StmtCtx.RCCheckTS:
+		stmtTSFuture = sessiontxn.ConstantFuture(p.latestOracleTS)
 	default:
-		stmtTSFuture = sessiontxn.NewOracleFuture(p.ctx, p.sctx, sessVars.TxnCtx.TxnScope)
+		stmtTSFuture = p.getOracleFuture()
 	}
 
 	p.stmtTSFuture = stmtTSFuture
 }
 
+func (p *PessimisticRCTxnContextProvider) getOracleFuture() sessiontxn.FuncFuture {
+	txnCtx := p.sctx.GetSessionVars().TxnCtx
+	future := sessiontxn.NewOracleFuture(p.ctx, p.sctx, txnCtx.TxnScope)
+	return func() (ts uint64, err error) {
+		if ts, err = future.Wait(); err != nil {
+			return
+		}
+		txnCtx.SetForUpdateTS(ts)
+		ts = txnCtx.GetForUpdateTS()
```

Comment:
When is it possible that the `ts` changes after setting and getting from txnCtx?

Reply:
Ts can be less than txnCtx.startTS, althought it is not possible.